library/std/src/sys/pal/unix/locks/pthread_mutex.rs

```diff
@@ -103,8 +104,17 @@ impl Mutex {
 
     #[inline]
     pub unsafe fn lock(&self) {
+        #[cold]
+        #[inline(never)]
+        fn fail() -> ! {
+            let error = Error::last_os_error();
+            panic!("failed to lock mutex: {error}");
+        }
+
         let r = libc::pthread_mutex_lock(raw(self));
-        debug_assert_eq!(r, 0);
+        if r != 0 {
```

Comment:
It might be worth explicitly noting in that comment that the unreachability _depends on_ the content of the `init` routine, and specifically the fact that it includes `libc::PTHREAD_MUTEX_NORMAL`. Were `init` to change, this would become reachable. (I like to point things like this out in comments if they can't be totally understood using local reasoning alone.)

Reply:
Sure, I've added it!